### ISSUE TITLE
Suggestion: upgradable build queue

### ISSUE THREAD

**Issue by Katorone** [ExileNG/NGSpaceCompany#21]
With this suggestion I hope to give some ideas about a possible build queue system.  Even if you don't think a build queue suits this game, I hope the ideas here can serve as inspiration in other ways.   Active players can use the queue to save up for big projects while manually buying other machines.  Idle players can use the queue to maximize their resource gains while they're away.  The queue would be initially unlocked through a dark matter upgrade.

- Level 1: 
  - Unlocked through Dark Matter
  - The queue has 3 slots for a single building assignment (machine, lab, research, dyson segment, ship, ...) and will build these items sequentially. 
  - If the Multi Buy upgrade is bought, the player can queue duplicates of a building in a single slot. (eg 50 dyson segments or 11 energy efficiency upgrades)
- Level 2:
  - Unlocked through science
  - The build queue gets a toggle to buy sequentially or to buy anything it can when resources are available
- Level 3-10
  - Unlocked & upgraded through science
  - The build queue expands with 1 slot per tier of this research
- Level 11
  - Unlocked through ultrite
  - A smarter queue that reserves needed materials.  Player has 4k metal, a queued machine requires 1k metal but doesn't have the other required resources yet.  This 1k metal gets reserved by the queue, the player can safely spend 3k metal on manual buildings.
- Level 12
  - Unlocked through ultrite
  - The smartest queue.  Not only does this queue reserve materials, it also takes into account resource gain while waiting on other resources.  Example; a building can be built in 2 hours when there will be enough meteorite, it also needs metal and the metal storage is full.  The player is allowed to spend metal until all resources for the queued building are on the same timer.  Eg 5 minutes for meteorite & 5 minutes for metal.

The biggest issue I see here is dealing with (auto) EMC.  For manual EMC a recalculation of the queue needs to trigger.  Auto EMC can be abstracted to a per second buff of a resource:  
`math.min(
  energy_gain * auto_emc_interval,
  total_battery_capacity
) / EMC_Factor / auto_emc_interval
`
While on a higher interval this can cause some inaccuracies, it would be at most a minute.  Still, this trade-off might be more interesting than recalculating the queue every time auto EMC triggers.  Instead only having to recalculate when a building gets bought manually or the auto EMC settings are changed.

Another possible issue is that players shouldn't be able to queue items that are impossible to build.  Specially when resources are reserved, this could make it impossible for them to build anything else, until manually removing the queued item or toggling the reservation.  An example of this is queuing a storage upgrade for lunarite when the metal storage is too small.  This would reserve the metal, making it impossible to buy the metal storage upgrade.
